tests/chainer_tests/test_variable.py

```diff
@@ -983,358 +981,325 @@ def v():
-    {'x_shape': ()},
-)
-class TestVariableToCpu(unittest.TestCase):
+class VariableToDeviceFamilyBase(object):
 
-    def setUp(self):
-        self.x = np.zeros(self.x_shape, dtype=np.float32)
-        self.gx = np.ones_like(self.x)
+    x_shape = None
+    x_with_grad = True
+    x_with_graph = True
 
-    def check_to_cpu(self, x, gx, requires_grad=True):
-        x_var = chainer.Variable(x, requires_grad=requires_grad)
+    def setup(self, src_backend_config):
+        x = np.zeros(self.x_shape, dtype=np.float32)
+        gx = np.ones_like(x)
 
-        set_grad_var = requires_grad or not isinstance(x, chainerx.ndarray)
-        if set_grad_var:
-            x_var.grad_var = chainer.Variable(gx, requires_grad=requires_grad)
+        x, gx = src_backend_config.get_array((x, gx))
 
-        x_var.to_cpu()
+        # *_src arrays are kept for the post transfer checks.
+        self.x_src = x
+        self.gx_src = gx
 
-        assert x_var.xp is np
-        assert x_var._has_chainerx_array is False
-        assert x_var.node is not None
-        assert isinstance(x_var.data, np.ndarray)
-        assert x.shape == x_var.shape
-        assert x.dtype == x_var.dtype
-        np.testing.assert_array_equal(
-            backend.CpuDevice().send(x_var.data), backend.CpuDevice().send(x))
-
-        if set_grad_var:
-            assert isinstance(x_var.grad, np.ndarray)
-            assert gx.shape == x_var.grad.shape
-            assert gx.dtype == x_var.grad.dtype
-            np.testing.assert_array_equal(
-                backend.CpuDevice().send(x_var.grad),
-                backend.CpuDevice().send(gx))
-            assert x_var.grad_var is not None
-            assert x_var.grad_var.node is not None
-        else:
-            assert x_var.grad is None
-            assert x_var.grad_var is None
-
-        orig_xp = backend.get_array_module(x, gx)
-        if orig_xp is np:
-            assert x_var.data is x
-            assert x_var.grad is gx
-        else:
-            assert x_var.data is not x
-            assert not set_grad_var or x_var.grad is not gx
-
-        assert x_var.xp is not chainerx
-        assert x_var._has_chainerx_array is False
-
-    def test_to_cpu_from_cpu(self):
-        self.check_to_cpu(self.x, self.gx)
-
-    @attr.gpu
-    def test_to_cpu_from_gpu(self):
-        self.check_to_cpu(cuda.to_gpu(self.x), cuda.to_gpu(self.gx))
-
-    @attr.chainerx
-    def test_to_cpu_from_chainerx(self):
-        self.check_to_cpu(
-            chainerx.array(self.x),
-            chainerx.array(self.gx),
-            requires_grad=False)
+        x = chainer.Variable(x)
+        gx = chainer.Variable(gx)
 
-    @attr.chainerx
-    def test_to_cpu_from_chainerx_requiring_grad(self):
-        with self.assertRaises(RuntimeError):
-            self.check_to_cpu(
-                chainerx.array(self.x),
-                chainerx.array(self.gx),
-                requires_grad=True)
+        if self.x_with_graph:
+            x *= x
```

Comment:
This is a bit difficult to understand. How about treating intel64 backend a special case, or at least leave some comment.

Reply:
Added a comment.

Reply:
Well, the original code was `x = 2 * x`. How about writing that? In-place operation is confusing here, in the first place.
By the way, Is "python scalar" essential? If `x = x * x` works, how about using that?

Reply:
Changed to a reassignment.

Reply:
I won't say that an example is essential. However, it makes the comment a lot clearer demonstrating to what concrete types `x` may be cast and under what circumstances.

Reply:
Hm, if `x = x * x` works, I don't even think the comment is necessary at all.
What do you think about that?

Reply:
I'm fine either way. I can remove it. I must say that this behavior was a surprise to me though.

Reply:
Removed.

Reply:
>I must say that this behavior was a surprise to me though.

Yes, but it's no longer relevant to this test code.